### PR TITLE
ping/rust: Refactor into multiple binaries

### LINKED ISSUE #71: Split Rust project into one binary per dependency version
This should hopefully be more maintainable down the line and help resolve errors such as https://github.com/libp2p/rust-libp2p/actions/runs/3390990380/jobs/5635708451.

### PR DESCRIPTION
**Pull request by thomaseizinger**
By having one Rust binary per version, we can vary the actual binary from version to version and f.e. fix deprecated API calls. It does introduce a bit of duplication between the different versions but I'd rather have that then not being able to adapt the tests to new APIs.

Instead of activating a feature per libp2p version, we add them all as dependencies. This ensures all transitive dependencies are properly tracked in `Cargo.lock`. Additionally, this gives us a single place we are can activate all the feature.

For `master` and pull-request builds, we replace the git target or rev with the one coming from the CI build. Once we trigger a build, `cargo` will update and resolve the necessary dependencies before that, thus fixing problems such as https://github.com/libp2p/rust-libp2p/pull/2972.

Fixes #71.

### CODE REVIEW

**Review by mxinden — COMMENTED** [review]
I like the direction this is taking. Thanks @thomaseizinger!

I will take a more in-depth look.

Mentioning it here to not forget, this pull request also needs to update the root `README.md`.

**Review by mxinden — APPROVED** [review]
Changes look good to me. Before we merge, @thomaseizinger can you confirm that this fixes https://github.com/libp2p/rust-libp2p/pull/2972/?

File: ping/_compositions/go-rust-interop-latest.toml
Comment (mxinden) [inline ping/_compositions/go-rust-interop-latest.toml:95]: :tada:

**Comment by thomaseizinger** [comment]
> Changes look good to me. Before we merge, @thomaseizinger can you confirm that this fixes https://github.com/libp2p/rust-libp2p/pull/2972/?

I'll have to check that locally!

File: ping/_compositions/go-rust-interop-latest.toml
Comment (thomaseizinger) [reply ping/_compositions/go-rust-interop-latest.toml:95]: Unfortunately, this doesn't seem to work for Git dependencies that point to something else other than the master branch.

**Comment by thomaseizinger** [comment]
> > Changes look good to me. Before we merge, @thomaseizinger can you confirm that this fixes [libp2p/rust-libp2p#2972](https://github.com/libp2p/rust-libp2p/pull/2972)?
> 
> I'll have to check that locally!

I had to make several changes to make all the usecases work but I've verified it locally now and it all works. I made some more changes to the Dockerfile that should hopefully make it more maintainable.

Pretty much all the configuration is in `Cargo.toml` now. It behaves like a normal cargo project and you can verify the build easily locally.

When we release a new version, we only need to append an entry to `[[groups]]` and add a new binary with the next version.

**Comment by thomaseizinger** [comment]
We should also get much better build-performance out of the new Dockerfile, assuming testground has enough space to cache our docker layers (spoiler: they are massive, ~ 20GB).

All released versions should always be pre-build and not require and changes.
To speed up builds for master and PRs, we can occasionally bump the revision that is in the manifest file to a more recent one.

**Comment by thomaseizinger** [comment]
I don't understand why the build is failing: https://github.com/libp2p/test-plans/actions/runs/3468917140/jobs/5795285685#step:6:326

For some reason, testground is building an old Dockerfile.

File: ping/rust/Dockerfile
Comment (mxinden) [inline ping/rust/Dockerfile:11]: As an aside, you can use a different delimiter than `/` with `sed`, e.g. `#`.

See https://stackoverflow.com/a/5864155/4011134

**Comment by mxinden** [comment]
> I don't understand why the build is failing: https://github.com/libp2p/test-plans/actions/runs/3468917140/jobs/5795285685#step:6:326
> 
> For some reason, testground is building an old Dockerfile.

Just a thought: Is the environment variable properly expanded, or accidentally used in plaintext?

```
Updating git repository `[https://github.com/libp2p/rust-libp2p`](https://github.com/libp2p/rust-libp2p%60)
error: failed to get `libp2p` as a dependency of package `testplan v0.1.0 (/usr/src/testplan/plan)`

Caused by:
  failed to load source for dependency `libp2p`

Caused by:
  Unable to update https://github.com/libp2p/rust-libp2p?rev=$.Env.GitTarget
  ```

File: ping/rust/Dockerfile
Comment (thomaseizinger) [reply ping/rust/Dockerfile:11]: My god, knowing this would have saved me quite some time 😂

I'll try to simplify that then.

**Comment by thomaseizinger** [comment]
> > I don't understand why the build is failing: https://github.com/libp2p/test-plans/actions/runs/3468917140/jobs/5795285685#step:6:326
> > 
> > For some reason, testground is building an old Dockerfile.
> 
> Just a thought: Is the environment variable properly expanded, or accidentally used in plaintext?
> 
> ```
> Updating git repository `[https://github.com/libp2p/rust-libp2p`](https://github.com/libp2p/rust-libp2p%60)
> error: failed to get `libp2p` as a dependency of package `testplan v0.1.0 (/usr/src/testplan/plan)`
> 
> Caused by:
>   failed to load source for dependency `libp2p`
> 
> Caused by:
>   Unable to update https://github.com/libp2p/rust-libp2p?rev=$.Env.GitTarget
>   ```

It was when I tested it locally!

File: ping/rust/Dockerfile
Comment (thomaseizinger) [inline ping/rust/Dockerfile:23]:
```suggestion
COPY --from=builder /usr/src/testplan/plan/target/debug/${BINARY_NAME} /usr/local/bin/
RUN mv /usr/local/bin/${BINARY_NAME} /usrl/local/bin/testplan
```

File: ping/rust/Dockerfile
Comment (thomaseizinger) [inline ping/rust/Dockerfile:26]:
```suggestion
ENTRYPOINT ["testplan"]
```

File: ping/rust/Dockerfile
Comment (mxinden) [inline ping/rust/Dockerfile:23]:
```suggestion
RUN mv /usr/local/bin/${BINARY_NAME} /usr/local/bin/testplan
```

CI failures:

```
Step 17/20 : RUN mv /usr/local/bin/${BINARY_NAME} /usrl/local/bin/testplan
 ---> Running in e577b39f88fe
mv: cannot move '/usr/local/bin/' to '/usrl/local/bin/testplan': No such file or directory
```

File: ping/rust/Dockerfile
Comment (mxinden) [inline ping/rust/Dockerfile:23]: Is `BINARY_NAME` defined here?

CI error:
```
Step 17/20 : RUN mv /usr/local/bin/${BINARY_NAME} /usr/local/bin/testplan
 ---> Running in 66b2252d5e55
mv: cannot move '/usr/local/bin/' to a subdirectory of itself, '/usr/local/bin/testplan'
```

**Review by mxinden — APPROVED** [review]

File: ping/rust/Dockerfile
Comment (thomaseizinger) [reply ping/rust/Dockerfile:23]: Oh my god! The ARG definition from above doesn't carry over! That is the bug! Thanks haha!

Comment (thomaseizinger) [inline ping/rust/Dockerfile:23]:
```suggestion
ARG BINARY_NAME
COPY --from=builder /usr/src/testplan/plan/target/debug/${BINARY_NAME} /usr/local/bin/${BINARY_NAME}
```

File: ping/rust/Dockerfile
Comment (thomaseizinger) [inline ping/rust/Dockerfile:19]:
```suggestion
```

**Comment by thomaseizinger** [comment]
Oh my god it is finally working! :tada:

File: ping/rust/Dockerfile
Comment (thomaseizinger) [inline ping/rust/Dockerfile:21]:
```suggestion
# Build the requested binary: Cargo will update the lockfile on changed manifest (i.e. if one of the above `sed`s patched it).
ARG BINARY_NAME
RUN cd ./plan/ \
    && cargo build --bin=${BINARY_NAME} \
    && mv /usr/src/testplan/plan/target/debug/${BINARY_NAME} /usr/src/testplan/plan/target/debug/testplan
```

**Comment by thomaseizinger** [comment]
Okay, this is ready to go in from my end!

Tagging other maintainers for visibility. cc @jxs @elenaf9

**Comment by thomaseizinger** [comment]
It seems that the initial docker layers are not cached between CI runs: https://github.com/libp2p/test-plans/actions/runs/3484164725/jobs/5828413991#step:6:46

I think that was the case before as well though, so this doesn't make things worse I think.

File: ping/rust/Dockerfile
Comment (mxinden) [inline ping/rust/Dockerfile:4]: @laurentsenta tagging you here for visibility, given that you have worked on the caching optimizations below.

With this pull request `cargo build` is the same across rust-libp2p versions (apart from `master` and a specific git commit hash), i.e. it builds all rust-libp2p versions and can thus be reused when testing different rust-libp2p versions. This results in the below simplifications.

Let us know in case you have any objections.

Comment (laurentsenta) [reply ping/rust/Dockerfile:4]: @mxinden thanks for the heads up; no objection on my side, but a question and a comment:

(q) it seems like we "just" moved the version indirection from feature flags to one layer up, but no big change in the `Cargo.toml`:

So we replace `cargo build --features="${VERSION_ID}"` with `cargo build --bin=${VERSION_ID}`

Why is this improving how we track dependencies in the `Cargo.lock`?

(c) We're dropping dependency caching, i.e. "do not download and rebuild all my dependencies when I change one line in my main.rs".

I added this to make the experience of iterating on the test code more bearable. We might want to add back that feature eventually (through dummy mains, `cargo-chef`, or something else).

File: ping/_compositions/rust.toml
Comment (jxs) [inline ping/_compositions/rust.toml:20]: should here also be the `46.1`?

File: ping/_compositions/rust.toml
Comment (jxs) [inline ping/_compositions/rust.toml:25]:
```suggestion
BinaryName = 'testplan_0451'
```

File: ping/rust/src/bin/testplan_0490.rs
Comment (jxs) [inline ping/rust/src/bin/testplan_0490.rs:19]: we are not using the `derive` macro here, do we need `#[allow(deprecated)]` ?

**Review by jxs — APPROVED** [review]
LGTM Thomas :muscle:

File: ping/rust/src/bin/testplan_0490.rs
Comment (thomaseizinger) [reply ping/rust/src/bin/testplan_0490.rs:19]: It is still a deprecated function!

File: ping/_compositions/rust.toml
Comment (thomaseizinger) [reply ping/_compositions/rust.toml:20]: Not sure, this is take verbatim from current master.

File: ping/_compositions/rust.toml
Comment (thomaseizinger) [reply ping/_compositions/rust.toml:25]: Can you send a follow-up if you care much? :)

I'd prefer to merge this asap to unblock https://github.com/libp2p/rust-libp2p/pull/2972.

File: ping/rust/Dockerfile
Comment (laurentsenta) [reply ping/rust/Dockerfile:4]: Got it; thanks for the clarification and for updating the description

> If somehow possible, I'd like to avoid containers for local iteration. Can testground run local binaries too?

Yes, it doesn't do network isolation and simulation, so you might not be able to use this as the main workflow to develop & fix tests.

> What is important is splitting the binaries per version to allow fine granular control over the code per version.

One last comment on this, if that's the point of this PR, is it possible we're trading "simple" for "easy" and we'll pay in the long term?

We're replacing "I have one main, with many shims for backward compatibility" with "I have many mains re-implementing a shared API".

The go version has been around for longer, so we have more examples of this:

- [One shared main](https://github.com/libp2p/test-plans/blob/b97b36aa2754c77795635facbf96c8504bdd9135/ping/go/main.go)
- [Shim some parts of the library](https://github.com/libp2p/test-plans/blob/b97b36aa2754c77795635facbf96c8504bdd9135/ping/go/compat/libp2p.v0.20.go#L1-L2)  (build flags in the header are ~= rust feature flags)

Note how surgical we can be when we support new versions, upgrading from v0.20 to v0.21 is "just" adding a flag because there was no API-breaking change.

With the new approach, it seems like we have 7 main files, ~600 lines, and only ~40 lines that are different.

Comment (mxinden) [reply ping/rust/Dockerfile:4]: > One last comment on this, if that's the point of this PR, is it possible we're trading "simple" for "easy" and we'll pay in the long term?

I see your point. Though unfortunately I don't see how we can support future breaking changes in rust-libp2p with the current "simple" approach. In other words, I don't see an alternative to the "easy" approach going forward.

All that said, I think it is relatively easy to iterate on this part of the code base. In case we come up with a better design, let's refactor `ping/rust`.

> > (c) We're dropping dependency caching, i.e. "do not download and rebuild all my dependencies when I change one line in my main.rs".
> > I added this to make the experience of iterating on the test code more bearable. We might want to add back that feature eventually (through dummy mains, `cargo-chef`, or something else).
> 
> Yeah I've been there. I am okay with it being added back. I stripped it to simplify things as much as possible during debugging.

@thomaseizinger would you mind adding it back in? I suggest not blocking this pull request, but instead do so in a follow-up pull request.

Comment (thomaseizinger) [reply ping/rust/Dockerfile:4]: > > Yeah I've been there. I am okay with it being added back. I stripped it to simplify things as much as possible during debugging.
> 
> @thomaseizinger would you mind adding it back in? I suggest not blocking this pull request, but instead do so in a follow-up pull request.

I'll have a look. Right now, I am just happy to be unblocked on my PR :)

**Comment by thomaseizinger** [comment]
Any sed experts here?

Something doesn't seem to quite work: https://github.com/libp2p/rust-libp2p/actions/runs/3499746591/jobs/5861640925

**Pull request merged**
